Read invitation parameters from wiki page

BenignBeppe opened the pull request:
Max age and min edits to qualify for invitation are defined in
hb_config. Values can be read from wiki pages.

BenignBeppe committed:
Read invitation parameters from wiki page

Max age and min edits to qualify for invitation are defined in
hb_config. Values can be read from wiki pages.

lokal-profil commented on hb_queries.py:
```diff
@@ -36,22 +37,26 @@ def make_teahouse_experiment_newbies(self):
 WHERE user_registration > DATE_FORMAT(
-    DATE_SUB(NOW(),INTERVAL 2 DAY),'%Y%m%d%H%i%s')
-AND user_editcount >=5
+    DATE_SUB(NOW(),INTERVAL {max_user_age} DAY),'%Y%m%d%H%i%s')
```
do we want to force formatting on this and the next parameter?

BenignBeppe replied:
Seems like a good thing to check. I'll add this.

lokal-profil commented on page_reader.py:
```diff
@@ -0,0 +1,73 @@
+            hb_config.apiurl,
+            data={
+                'action': "query",
+                'titles': page,
```
I believe you might have to do page.encode('utf-8') here.

But you can check with a unicode page name to make sure.

BenignBeppe replied:
Amazingly enough, character encoding doesn't seem to cause any trouble here. I did quick test and "User:FikarummetBot/Å" worked, both as `str` and `unicode`. I did changed the call to `get()´.

lokal-profil replied:
Great. I guess my worry stems from pre-`requests` days =)

lokal-profil commented on page_reader.py:
```diff
@@ -0,0 +1,73 @@
+    multiple requests.
+    """
+
+    if page in pages:
```
You should standardise the pagename first (capitalize first letter and either all spaces as underscores or all underscores as pages) to ensure the cached one is always found.

BenignBeppe replied:
Added function for normalizing page title.

lokal-profil commented on page_reader.py:
```diff
@@ -0,0 +1,73 @@
+    """Read a new line separated list from a page."""
+
+    page_content = _read_page(page)
+    list_ = [i for i in page_content.split("\n")]
```
Is the loop only there to get rid of the last empty line or do you also expect more empty lines?

Wondering since in the first scenario `page_content.strip().split("\n")` should work.

BenignBeppe replied:
I'm honestly not sure why I wrote it like this. Just `page_content.split("\n")` works, so I changed to this. Trailing new lines seems to be stripped before the content is received, as well.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -18,17 +18,22 @@
-import random
+import page_reader
+
+RE_USER_NAME = re.compile(r"\[\[User:(.*?)(\]\]|\|)")
```
"User:" depends on the language of the wiki. E.g. sv.wiki you can write either "User:" or "Användare:". Sadly for us signatures default to using the local language version.

So we would need `RE_USER_NAME = re.compile(r"\[\[(User|Användare):(.*?)(\]\]|\|)")` here and `RE_USER_NAME.match(wikitext).group(2)` below.

Of course it would be better to load the local language user-namespace from the config. (We don't need to load the english one since that is always allowed)

BenignBeppe replied:
User namespace is now defined in `hb_config`. "User" is always valid.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -78,8 +93,10 @@ def inviteGuests(prof, message_text, inviter):
     else:
         pass
 #     inviters = params['inviters'] #for TWA
+    page_list = page_reader.read_list_page("User:FikarummetBot/Inviters")
```
This pagename should be in the config (and sample).

I would even say that the config should just return a clean list here and deciding whether to load from page or from a provided list in the config (and possibly also any username from wikitext extraction) should be handled before.

BenignBeppe replied:
I moved the inviter list to `hb_config`.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -78,8 +93,10 @@ def inviteGuests(prof, message_text, inviter):
 #     inviters = params['inviters'] #for TWA
+    page_list = page_reader.read_list_page("User:FikarummetBot/Inviters")
+    inviter_list = [user_name_from_wikitext(i) for i in page_list]
     inviters = getEligibleInviters(elig_check, params['inviters'])
```
This should take `inviter_list` rather than `params['inviters']`

BenignBeppe replied:
Done.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -18,17 +18,22 @@
     return eligible_inviters
 
-def getEligibleInvitees(elig_check, potential_invitees, skip_templates):
+def getEligibleInvitees(elig_check, potential_invitees):
```
Might be worth adding a comment on line 79 so that we remember that we are in fact still sending `skip templates` (and `output namespace`) to the underlying function.

BenignBeppe replied:
Done.

lokal-profil reviewed (changes requested):
Good in general. A few minor changes though.

Also if you can drop the now unneeded entries from `hb_output_settings` it should be easier to spot which are left to be addressed.

BenignBeppe committed:
Implemented review comments.

BenignBeppe commented:
I've commented out the parameters no longer used in `hb_output_settings`. Of the remaining, skipped templates and user talk name space should be relatively straight forward. 'select query' may take some investigation to figure out what it actually does.

lokal-profil commented on page_reader.py:
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
+def _normalize_title(title):
+    """Normalize a page title.
+
+    If the title contains a colon, it is assumed that it is a namspace
```
typo namespace

BenignBeppe replied:
Fixed.

lokal-profil commented on page_reader.py:
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
+    """
+
+    parts = title.partition(":")
+    if parts[1] != "":
```
You can just do if parts[1]

lokal-profil replied:
see comment below

lokal-profil commented on page_reader.py:
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
+    if parts[1] != "":
+        # Since a colon was found, this should contain a namespace.
+        namespace, _, name = parts
+        capitalized_name = name[0].upper() + name[1:]
```
You should also capitalize the namespace `user:lisa == `User:Lisa`

lokal-profil replied:
see comment below

lokal-profil commented on page_reader.py:
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
+        capitalized_title = "{}:{}".format(namespace, capitalized_name)
+    else:
+        name = parts[0]
+        capitalized_name = name[0].upper() + name[1:]
```
should be capitalized_title in order to get picked up by normalized_title below

lokal-profil replied:
see comment below

lokal-profil commented on page_reader.py:
```diff
@@ -65,9 +68,28 @@ def _read_page(page):
+        name = parts[0]
+        capitalized_name = name[0].upper() + name[1:]
+    normalized_title = capitalized_title.replace("_", " ")
+    return normalized_title
```
Could this (i.e from partition to return) not be done as
```
namespace, _, name = title.partition(":")
if name:
    namespace = namespace[0].upper() + namespace[1:]
else:
    name = namespace
    namespace = ''
name = name[0].upper() + name[1:]
capitalized_title = "{}{}{}".format(namespace, _, name)
normalized...
```

BenignBeppe replied:
Yep, that's probably easier to read. I added it with a couple of minor changes.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -25,8 +25,13 @@
+# Build an RE pattern that accepts None as user namespace to just use
+# the default "User".
+RE_USER_NAME = re.compile(r"\[\[(User{:s}):(.*?)(\]\]|\|)".format(
+    ("|" + str(hb_config.user_namespace)) * (hb_config.user_namespace != None)
```
I believe "is not" is the more common test for None comparison

BenignBeppe replied:
Done.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -71,11 +76,13 @@ def user_name_from_wikitext(wikitext):
     """
 
-    return RE_USER_NAME.match(wikitext).group(1)
+    return RE_USER_NAME.match(wikitext).group(2)
```
I think we need a try/except block here since (when I quickly test it) I get AttributeError if the input is weirdly formatted. Since we are reading this from a wikipage I suggest adding the robustness of simply skipping any such. Easiest is to simply return `None` on (that) exception and add `inviter_list = filter(None, inviter_list)` later.

BenignBeppe replied:
Now, only valid user names (from links) are added. When an entry doesn't match, a debug message is written to stdout.

lokal-profil commented on scripts/teahouse/send_th_invites.py:
```diff
@@ -93,9 +100,8 @@ def user_name_from_wikitext(wikitext):
-    page_list = page_reader.read_list_page("User:FikarummetBot/Inviters")
-    inviter_list = [user_name_from_wikitext(i) for i in page_list]
-    inviters = getEligibleInviters(elig_check, params['inviters'])
+    inviter_list = [user_name_from_wikitext(i) for i in hb_config.inviters]
```
could you add this variable to the config sample file?

BenignBeppe replied:
Added.

lokal-profil reviewed (changes requested):
comments on _normalize_title and RE_USER_NAME

BenignBeppe committed:
While fixing the review comments, page_reader was reworked a bit. It now
contains a few more variables (beyond page cache) that are set the
first time they are used.

Also removed the shebang from the config.

lokal-profil commented on hb_config.sample.py:
```diff
@@ -51,3 +51,7 @@
 summary = '{invitee:s}, you are invited to the Teahouse!'
+
+# The user that can be mentioned in `message` above. A user will be
+# selected at random from this list, for each message.
```
Maybe mention page_reader.read_user_list_page() here.

BenignBeppe committed:
Added comment for page_reader.read_user_list_page().
